Navigation block: use `<dialog>`

stokesman opened the pull request:
## What?
Alternative to #69665
Closes #69635

Utilizes the `<dialog>` element to wrap the navigation content and invokes it as a modal for the overlay.

## Why?
- This ensures the overlay’s layout is not unexpectedly contained by ancestors
- Prepares the way for the overlay to function without JS

## How?
- Replaces a `<div>` with `<dialog>`
- Uses Invoker `command` and `commandfor` attributes on the buttons to open and close the dialog
- Feature detects support of Invoker commands and if lacking uses JS to open and close the dialog
- Removes implementation of “focus trapping” in the dialog
- In the editor, adds code to prevent root/document scrolling while the dialog is open (this is outside the focus of this PR and could be dropped)

## Testing Instructions
#69635 is an issue that is reproducible in Firefox or Safari so test with one of those browsers.

Here is markup for testing, though you’ll have to update the navigation for your site.
```html
<!-- wp:cover {"url":"https://pd.w.org/2021/05/97060b09513e73a89.38852584-2048x1536.jpeg","dimRatio":50,"customOverlayColor":"#001240","isUserOverlayColor":true,"minHeight":20,"minHeightUnit":"vw","sizeSlug":"full","style":{"border":{"radius":"99em","color":"#0069e9","width":"6px"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-cover has-border-color" style="border-color:#0069e9;border-width:6px;border-radius:99em;min-height:20vw"><img class="wp-block-cover__image-background size-full" alt="" src="https://pd.w.org/2021/05/97060b09513e73a89.38852584-2048x1536.jpeg" data-object-fit="cover"/><span aria-hidden="true" class="wp-block-cover__background has-background-dim" style="background-color:#001240"></span><div class="wp-block-cover__inner-container"><!-- wp:group {"style":{"position":{"type":"sticky","top":"0px"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:navigation {"ref":4} /--></div>
<!-- /wp:group --></div></div>
<!-- /wp:cover -->
```

1. Open a post
2. Paste in the markup from above and update the navigation
3. Switch to "Mobile" device preview
4. Open the navigation menu inside the Cover block
5. Verify the overlaid menu appears as expected
6. View/Preview the post
7. Ensure the viewport is narrow enough that the navigation is in mobile/overlay mode
8. Open the navigation menu
9. Verify the overlaid menu appears as expected

## Screenshots or screencast

### Editor
|Before|After|
|-|-|
|<video src="https://github.com/user-attachments/assets/afe87a27-9f8d-416f-8543-91a571fefe59"/>| <video src="https://github.com/user-attachments/assets/ae5c3407-6a4d-4f1f-b874-2640c8f6566d"/> |

### Front end
|Before|After|
|-|-|
|<video src="https://github.com/user-attachments/assets/e8478adc-1d8c-4991-acd2-6889db9001d9"/>|<video src="https://github.com/user-attachments/assets/fc076619-79dd-429b-85d5-e3b4d2bea0ed"/>|

stokesman committed:
Navigation block: use <dialog>

stokesman committed:
Restore `Escape` handling to close sub menus

stokesman committed:
Ensure closing logic is applied by listening to dialog `close`

Also, simplify `closeMenu` and its usage

stokesman committed:
Temporarily omit focus trap testing in E2E

stokesman committed:
Try fixing color inheritance due to UA styles on <dialog>

stokesman committed:
Avoid closing submenu on hover out if opened by click

stokesman committed:
Use async directive for submenu keydown

stokesman committed:
Use Invoker Commands for dialog

stokesman committed:
Consolidate open/close effectual code

Also:
- Add `modal` to nav element’s context from PHP
- Rename `initMenu` to `effectOpenClose`

stokesman committed:
Remove tabbing assertions; remove webkit specific test

westonruter commented on packages/block-library/src/navigation/edit/responsive-wrapper.js:
```diff
@@ -90,16 +127,20 @@ export default function ResponsiveWrapper( {
-					<div { ...dialogProps }>
+					<div
+						className="wp-block-navigation__responsive-dialog"
+						aria-label={ isOpen && __( 'Menu' ) }
```
I learned recently that `aria-label` doesn't do anything on elements that don't have a `role`: https://gomakethings.com/aria-label-requires-a-valid-role/

afercia replied:
That's correct but it's worth reminding the `<dialog>` element does have e native role. As such, it is exposed in the browser's accessibility tree as `role=dialog`. Adding an aria-label does work as expected. Screenshot illustrating how manually adding an aria-label on the MDN demo does work:

![Screenshot 2025-03-30 at 13 58 52](https://github.com/user-attachments/assets/164d7275-4214-45ee-9cf8-b36c5d03e489)

stokesman replied:
Thanks for highlighting this. The attribute was already here but I did move the code for it—it was [here](https://github.com/WordPress/gutenberg/pull/69720/commits/3ea96bfabf648ce8445e3ce3cb677754cc0d06c9#diff-098771364317caf1591592ddeced4a2c1fcae9af8806dd29b6f6d537642e1fcdL69-L77). Now this element is wrapped in the `dialog` so I’d removed the `role` and `aria-modal` attributes. I suppose this label should move to the `dialog` element.

stokesman replied:
I moved the `aria-label` to the `dialog` element in 01643fad63c325dba3b88227b512c6845ac13ee4. This could still use some scrutiny. The label is applied only when the overlay is open. That’s how it was yet now the overlay can open without JS and the label won’t be applied. I think the idea behind conditionally applying it is to avoid its presence when the navigation has no overlay—i.e. in larger viewports. With this PR, the `dialog` element is there regardless of the viewport size and perhaps it’s not wrong for it to have the label even when it’s not acting as a modal.

westonruter commented:
Question: In https://github.com/WordPress/gutenberg/pull/69665#issuecomment-2756729651 you said you started this branch from my PR?

Just curious because I don't see my commit in this PR's history.

westonruter commented on packages/block-library/src/navigation/view.js:
```diff
@@ -155,54 +116,50 @@ const { state, actions } = store(
+				}
+				if ( ctx.type === 'overlay' ) {
+					// Toggles `has-modal-open` class on the <html> root.
+					document.documentElement.classList.toggle(
+						'has-modal-open',
+						state.isMenuOpen
+					);
```
I think it would be ideal if this class toggling wasn't required. Namely, the CSS rule could check if `body:has(dialog.wp-block-navigation__responsive-container:open)` instead. Then this would not require any JS to apply the styling.

westonruter replied:
Namely:

https://github.com/WordPress/gutenberg/blob/a6932935ae7072e59b5677c1d848476d978c69f3/packages/block-library/src/navigation/style.scss#L702-L711

and

https://github.com/WordPress/gutenberg/blob/a6932935ae7072e59b5677c1d848476d978c69f3/packages/block-library/src/navigation/style.scss#L731-L735

`html.has-modal-open` can be replaced with `html:has(dialog.wp-block-navigation__responsive-container:open)`.

And I suppose `.has-modal-open &` can be replaced with `html:has(dialog.wp-block-navigation__responsive-container:open) &`

stokesman replied:
> I think it would be ideal if this class toggling wasn't required.

I’d agree yet I kept the classes in accord [this part](https://developer.wordpress.org/block-editor/contributors/code/backward-compatibility/#class-names-and-dom-updates) of the BC guidelines:
> Class names and DOM nodes used inside the tree of React components are not considered part of the public API and can be modified.

> Changes to these should be done with caution as it can affect the styling and behavior of third-party code (Even if they should not rely on these in the first place). Keep the old ones if possible. If not, document the changes and write a dev note.

Thanks to your suggestions though I realized I missed adding a no-JS equivalent selector for this one:
https://github.com/WordPress/gutenberg/blob/a6932935ae7072e59b5677c1d848476d978c69f3/packages/block-library/src/navigation/style.scss#L706

westonruter replied:
Ah, got it. Ok, I think we should then keep the class name for now, but we can add a comment to say it is deprecated, and we can write a dev note to comment this. I'll do a search for themes in the directory to see if anyone is targeting this class name so we'll have a list for outreach.

westonruter replied:
There are 132 themes that are using this class name, at least at the time that WPdirectory last indexed the directory: https://wpdirectory.net/search/01JQQ5VY70TAAEMSXTAQVE69GW

stokesman replied:
Even were these class names to be kept indefinitely it would be nice to do outreach that recommends updating any selectors using them because otherwise the theme’s styles won’t apply if JS is unavailable.

stokesman commented on test/e2e/specs/editor/blocks/navigation-frontend-interactivity.spec.js:
```diff
@@ -66,10 +66,14 @@ test.describe( 'Navigation block - Frontend interactivity', () => {
-			await expect( closeMenuButton ).toBeFocused();
-			await pageUtils.pressKeys( 'Shift+Tab', { times: 2, delay: 50 } );
-			await expect( overlayMenuFirstElement ).toBeFocused();
+			/**
+			 * @todo Determine if focus trapping is required. If not, these assertions
+			 * can be removed and seemingly the next test too — see its comment.
+			 */
+			// await pageUtils.pressKeys( 'Tab', { times: 2, delay: 50 } );
+			// await expect( closeMenuButton ).toBeFocused();
+			// await pageUtils.pressKeys( 'Shift+Tab', { times: 2, delay: 50 } );
+			// await expect( overlayMenuFirstElement ).toBeFocused();
```
This tests focus wrapping, i.e. the focus moving either to the first tabbable to the last or vice versa. Since this PR removes the logic for this, focus doesn’t wrap (and tends to move into the browser’s UI). Yet thanks to the modal `dialog` nothing in the document outside the modal can be focused. I think that suffices and seems more accessible because focus isn’t trapped within the document by wrapping it. I just wanted to highlight this in case someone thinks differently.

I’ll go ahead and push a commit that removes these commented lines.

stokesman committed:
Revise CSS selector and add no-JS equivalent

stokesman commented:
> you said you started this branch from my PR?

I did but ended up squashing the commit. I’d done it before hearing from you and mostly because my first commit revised almost everything so it seemed a diff of little interest.

stokesman committed:
Move aria-label to dialog element

stokesman committed:
Revise submenu hover conditionals

stokesman commented on packages/block-library/src/navigation/view.js:
```diff
@@ -49,25 +49,18 @@ const { state, actions } = store(
-					// Only close on hover if the overlay is closed.
-					Object.values( overlayOpenedBy || {} ).filter( Boolean )
-						.length === 0
+					Object.values( overlayOpenedBy || {} ).some( Boolean )
```
This isn’t necessary but should introduce no behavioral differences. The `type === 'submenu'` conditions were useless because these handlers are only invoked by submenus (not the overlay). The `.filter( Boolean ).length === 0` is basically a more verbose version of `.some( Boolean )` which replaces it. Other than that this just revises the comments a bit.

westonruter commented:
See also @felixarntz's https://github.com/WordPress/gutenberg/pull/63917#issuecomment-2766592783:

> There's also #68755, so we may want to compare the approaches and scope of the changes between that and #69720.
> 
> In my PR (which admittedly is so far just experimental), I'm using popover for both the navigation overlay (by default only used on mobile) and for submenus (across all viewports).
> 
> One thing to call out is that unfortunately the Interactivity API implementation for the overlay and submenus is very intertwined - maybe this was done to reduce some code to write, but to me that doesn't look like a good choice because there are so many subtle nuances in different behavior. And now it makes decoupling more painful than it should be.
> 
> This is partly the reason why I think we need to consider changing both the overlay and the submenus implementation together. Because if you only change one of them to use popover or dialog, most of the Interactivity API code will still stick around for the other one.

stokesman committed:
Restore important rules for inheriting colors when overlay is not open

Also add e2e coverage since it was lacking

stokesman commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -728,28 +726,17 @@ button.wp-block-navigation-item__content {
 .wp-block-navigation__responsive-dialog {
 	position: relative;
 }
 
-// Adjust open dialog top margin when admin-bar is visible.
-// Needs to be scoped to .is-menu-open, or it will shift the position of any other navigations that may be present.
-.has-modal-open .admin-bar .is-menu-open .wp-block-navigation__responsive-dialog {
-	margin-top: $admin-bar-height-big;
-
-	// Handle smaller admin-bar.
-	@include break-medium() {
-		margin-top: $admin-bar-height;
-	}
-}
```
The overlay (`dialog`) is now in the top-layer and the admin bar is unable to appear in front of it so these styles are irrelevant.

stokesman commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -488,6 +488,13 @@ button.wp-block-navigation-item__content {
 		align-items: var(--navigation-layout-align, initial);
 	}
 
+	// If the responsive wrapper is present but overlay is not open, overlay colors
+	// shouldn't apply and !important is to override color classes or inline styles.
+	&:not(.is-menu-open) {
+		color: inherit !important;
+		background-color: inherit !important;
+	}
+
```
I had removed these because these properties were set to `inherit` otherwise and I didn’t realize the `!important` bit was key. I double-checked and found that out. Removing them hadn’t caused the e2e tests for colors to fail so along with restoring this I added an e2e test that fails without these.

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -498,10 +500,9 @@ button.wp-block-navigation-item__content {
 	// In a future version, we can explore more customizability.
-	&.is-menu-open {
-		display: flex; // Needs to be set to override "none".
+	&:is(.is-menu-open, [open]) {
```
Shouldn't this be `:open` as opposed to `[open]`? I [consulted](https://g.co/gemini/share/d4762d502566) Gemini who suggests they are functionally equivalent, but then says:

<blockquote><strong><code>:open</code> is often preferred:</strong> It's semantically clearer as it targets the <em>state</em> of the element (being open) rather than the implementation detail (having the <code>open</code> attribute). This aligns conceptually with other state pseudo-classes like <code>:checked</code>, <code>:disabled</code>, or <code>:focus</code>.</blockquote>

```suggestion
	&:is(.is-menu-open, :open) {
```

It's also 1 character shorter 😄

stokesman replied:
Thanks! I agree `:open` should be used. Done in 370b2046e0df76bf45a6dd12d8f0d6ab78359f80. Our style lint doesn’t like it but ignores work for now.

stokesman replied:
It turns out Safari doesn’t yet support `:open` and that was causing most of the recent e2e test failures. Even besides Safari the versions of Chrome and Firefox that don’t support it are still at usage levels above %1 which means we [shouldn’t use it yet](https://make.wordpress.org/core/handbook/best-practices/browser-support/#:~:text=Browsers%20with%20%3E%201%25%20usage%20based). So 966098f20b5f675502d0d24527dbb50bd3c1493c switches back to `[open]`.

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -612,7 +610,7 @@ button.wp-block-navigation-item__content {
 	@include break-small() {
 		&:not(.hidden-by-default) {
-			&:not(.is-menu-open) {
+			&:not(.is-menu-open, [open]) {
```
```suggestion
			&:not(.is-menu-open, :open) {
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -625,7 +623,7 @@ button.wp-block-navigation-item__content {
 		}
 
-		&.is-menu-open {
+		&:is(.is-menu-open, [open]) {
```
```suggestion
		&:is(.is-menu-open, :open) {
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -636,12 +634,12 @@ button.wp-block-navigation-item__content {
 // Default menu background and font color.
 .wp-block-navigation:not(.has-background)
-.wp-block-navigation__responsive-container.is-menu-open {
+.wp-block-navigation__responsive-container:is(.is-menu-open, [open]) {
```
```suggestion
.wp-block-navigation__responsive-container:is(.is-menu-open, :open) {
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -636,12 +634,12 @@ button.wp-block-navigation-item__content {
 
 .wp-block-navigation:not(.has-text-color)
-.wp-block-navigation__responsive-container.is-menu-open {
+.wp-block-navigation__responsive-container:is(.is-menu-open, [open]) {
```
```suggestion
.wp-block-navigation__responsive-container:is(.is-menu-open, :open) {
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -712,7 +710,7 @@ button.wp-block-navigation-item__content {
 	width: 100%;
 
-	.has-modal-open & {
+	:is(.is-menu-open, [open]) > & {
```
```suggestion
	:is(.is-menu-open, :open) > & {
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -728,28 +726,17 @@ button.wp-block-navigation-item__content {
-.is-menu-open .wp-block-navigation__responsive-close,
-.is-menu-open .wp-block-navigation__responsive-dialog,
-.is-menu-open .wp-block-navigation__responsive-container-content {
+.wp-block-navigation__responsive-container:is(.is-menu-open, [open])
```
```suggestion
.wp-block-navigation__responsive-container:is(.is-menu-open, :open)
```

westonruter commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -728,28 +726,17 @@ button.wp-block-navigation-item__content {
 // Prevent scrolling of the parent content when the modal is open.
-html.has-modal-open {
+html.has-modal-open,
+html:has(.wp-block-navigation__responsive-container[open]) {
```
```suggestion
html:has(.wp-block-navigation__responsive-container:open) {
```

westonruter committed:
Use `:open` instead of `[open]`

stokesman committed:
Add `:open` selector where it was missing

westonruter committed:
Set focus on first link when the nav menu opens

westonruter commented:
I just tested in Chrome 135 Dev Channel with the Experimental Web Platform features flag enabled (`#enable-experimental-web-platform-features`) so that invoker commands are available, and the mobile nav menu works!

JavaScript Enabled | JavaScript Disabled
--|--
<video src="https://github.com/user-attachments/assets/7d008e2b-a959-4e23-9134-76982edd2f48"> | <video src="https://github.com/user-attachments/assets/9176d47e-9a12-4ea4-ba92-cb307c23d590">

There is just a slight behavioral difference in where the keyboard focus in the `dialog` when it is opened. For the JS-enabled version, the focus is put on the initial nav men item while in the JS-disabled version the focus is put on the close button. I'm not sure which is better from an accessibility perspective. Is it better to have the focus be on the first nav menu item or to put it on the close button? If on the close button, then a user can quickly hit <kbd>ENTER</kbd>/<kbd>SPACE</kbd> again to close the menu. Otherwise, if the focus goes on the first nav nav menu item, they'd have to <kbd>SHIFT+TAB</kbd> to go back to the close button to close it. If it is truly better to focus on the first nav menu item, this is simple to implement just by adding the `autofocus` attribute to the first link which I've done in 42d6402af3a. This is something that @afercia has also recently discussed in https://github.com/WordPress/gutenberg/issues/69265#issuecomment-2684699629.

Something else I wonder about here: what happens in the case where the Interactivity API is delayed in initializing (e.g. due to a slow network) whereby the invoker command is used to open the nav menu first and then later once the Interactivity API initializes then the directives take over. I tried simulating this with the following patch:

```diff
diff --git a/packages/interactivity/src/index.ts b/packages/interactivity/src/index.ts
index b7d68fd2007..09700d50da3 100644
--- a/packages/interactivity/src/index.ts
+++ b/packages/interactivity/src/index.ts
@@ -57,5 +57,11 @@ export const privateApis = ( lock ): any => {
 	throw new Error( 'Forbidden access.' );
 };
 
-registerDirectives();
-init();
+document.addEventListener( 'click', () => {
+	setTimeout( () => {
+		console.info( 'Delayed Interactivity API init' );
+		registerDirectives();
+		init();
+	}, 1000 );
+}, { once: true } )
+
```

You can see in the following screencast that the nav menu opens with the invoker command first, and then the next time I open the nav menu (once the Interactivity API has initialized) it shows the tab focus is on the first nav menu item rather than the close button (as seen in the screencasts above, indicates that the JS code is now running):

[Screen recording 2025-04-01 11.32.06.webm](https://github.com/user-attachments/assets/58832a65-37f6-4cae-94ad-5b39a86027a6)

(I took this screencast before I committed 42d6402af3a to set the focus on the first nav menu item consistently.)

This doesn't necessarily seem like a problem to me, but I'm wondering if there is some potential for the DOM state to get out of sync with the Interactivity API state. It doesn't seem like this is currently causing a problem. However, if the browser supports invoker commands, should they be used exclusively instead of the Interactivity API? For example, there could be a `wp-init` directive which has an action that checks whether Invoker Commands are supported, and if so, it short-circuits and does nothing. Otherwise, if Invoker Commands aren't supported, then it can go ahead and add the necessary event listeners to open/close the nav menu with the store's actions. Once all browsers support invoker commands (maybe in a couple years), then the `wp-init` directive can be removed entirely.

westonruter commented on packages/block-library/src/navigation/index.php:
```diff
@@ -524,18 +518,28 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
 
+		// Set the autofocus attribute on the first link.
+		$tag_processor = new WP_HTML_Tag_Processor( $inner_blocks_html );
+		while ( $tag_processor->next_tag( array( 'tag_name' => 'A' ) ) ) {
```
This should also look for other focusable elements to correspond with:

https://github.com/WordPress/gutenberg/blob/42d6402af3a85b356390a453689b0adc9d45caf4/packages/block-library/src/navigation/view.js#L158-L165

I only implemented the first of these `focusableSelectors` so far:

https://github.com/WordPress/gutenberg/blob/42d6402af3a85b356390a453689b0adc9d45caf4/packages/block-library/src/navigation/view.js#L6-L14

westonruter commented:
> This sounds like it would be a nicer approach overall. Still, wouldn’t doing nothing when Invoker Commands are supported mean that the class names are omitted even with JS available? That could break theme styles and perhaps we’d want to hold off on that for the time being. I do find the idea/structure appealing so maybe it’s worth trying that direction now but not totally short-circuiting when Invoker Commands are supported.

Oh yes, good point. In that case, there is also the [`toggle` event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/toggle_event) which fires when the `dialog` is opened or closed. In the `wp-init` action, how about if Invoker Commands are supported, instead of just short-circuiting it could do this first:

```js
dialog.addEventListener( 'toggle', ( event ) => {
    document.documentElement.classList.toggle( 'has-modal-open', event.newState === 'open' );
    // And toggle any other "legacy" classes on other elements as well.
} );
```

stokesman commented:
That’s a good idea and I'm glad to learn the `toggle` event is meant to work for `dialog` too. Yet for now it looks like [Safari support is lacking](https://caniuse.com/mdn-api_htmlelement_toggle_event_dialog_elements) 😢. Still, I think roughly the same could be done using a `click` event just for the open button and the `close` event of the `dialog`.

stokesman committed:
Use [open] selector to support more browsers

stokesman committed:
Fix e2e `overlayMenu` attribute values

stokesman committed:
Update e2e test to work in webkit

afercia commented:
> There is just a slight behavioral difference in where the keyboard focus in the dialog when it is opened.

Based on my testing, the initial focus on a `<dialog>` element works a little differently across browsers especially when using the `autofocus` attribute.

The whole point of setting initial focus is to place it to an element that makes the most sense _based on the specific flow expected in that specific UI_. Howeer, sometimes there may be no specific flow expected. Two examples:

1. The 'Keyboard shortcuts' modal dialog contains a list of shortcuts. There's no action to take, no input to fill in. The dialog only contains two focusable elements: the Close button and the scrollable div. In this case, there's no specific flow the user is expected to follow. Not to mention a dialog may not contain any focusable element at all. It could be just text content. Where to put initial focus in these cases? The most reasonable place is the dialog element itself, which has a native role and is expected to be labeled. That's the reason why the Modal component provides a way to set initial focus on the container with role=dialog.
2. The 'Rename block' modal dialog does expect a specific user flow instead. There's only one action users can take: rename the block (well, they can also cancal). In this case, it makes sense to set initial focus to the 'name' input field.

The native behavior of the `<dialog>` element is no different. The browser attempts to identify a good candidate for the initial focus, with a fallback to the dialog element. Quoting:

https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element

An important part of user-facing dialog behavior is the placement of initial focus. The [dialog focusing steps](https://html.spec.whatwg.org/multipage/interactive-elements.html#dialog-focusing-steps) attempt to pick a good candidate for initial focus when a dialog is shown, but might not be a substitute for authors carefully thinking through the correct choice to match user expectations for a specific dialog. As such, authors should use the [autofocus](https://html.spec.whatwg.org/multipage/interaction.html#attr-fe-autofocus) attribute on the descendant element of the dialog that the user is expected to immediately interact with after the dialog opens. If there is no such element, then authors should use the autofocus attribute on the dialog element itself.

As such, given a generic component is not aware of the kind of content it will contain, I would recommend to set initial focus on the dialog element by default. Then, there should be a mehcanism to set the `autofocus` attribute on the most appropriate focusable element in the content, if any.

As mentioned, there's inconsistencies across browsers when using `autofocus` on the dialog element though. This:

`<dialog autofocus aria-label...`

Testing the major browsers, as of April 2, 2025 (macOS):
- Chrome 134: does _not_ set focus on the dialog element. It does on the first focusable element instead. This is against what the spec dictates.
- Firefox 137: does set focus on the dialog element. This is the expected behavior giving the code above.
- Safari 18.3.1: does set focus on the dialog element. This is the expected behavior giving the code above.
- Opera 117: does _not_ set focus on the dialog element. It does on the first focusable element instead. This is against what the spec dictates.

westonruter commented:
@afercia Thanks for the insights!

> As such, given a generic component is not aware of the kind of content it will contain, I would recommend to set initial focus on the dialog element by default. Then, there should be a mehcanism to set the `autofocus` attribute on the most appropriate focusable element in the content, if any.

I'm not sure I understand since the context here is the Navigation block's use of the `dialog` element. In the current JS-based implementation, the first nav menu item in the menu is focused when the menu is opened. As of 42d6402, this behavior is also replicated for `dialog` with declarative HTML, by adding the `autofocus` attribute to the first link. I tested your [Codepen](https://codepen.io/afercia/full/OPJrEEB) in Chrome 134, Firefox 137, and Safari 18.3.1 and in the second example ("Open a dialog with autofocus set on the second input field") results in the initial focus being placed on the second input field as expected. 

So isn't this PR going in the right direction by putting `autofocus` on the initial focusable item in the nav menu?

Nevertheless, I just found a complication here. When the nav menu is _not_ set to have the "overlay menu" always displayed (i.e. only use the overlay menu on mobile), the result is I see the first nav menu item is autofocused when I load the page on desktop:

> ![image](https://github.com/user-attachments/assets/980205c6-84c8-4ff1-897f-dbb2cd6a2fda)

This is clearly not right. This is due to the fact that the non-overlay menu is still being displayed inside of a `dialog` element which is forcibly being displayed via this CSS rule:

```css
@media (min-width:600px){
  .wp-block-navigation__responsive-container:not(.hidden-by-default):not(.is-menu-open,:open){
    background-color:inherit;
    display:block;
    position:relative;
    width:100%;
    z-index:auto;
  }
}
```

I wonder now if this is ideal? Would this be confusing from an accessibility perspective to have an element with a `dialog` role which is traversable/shown by default? 

What if, when the overlay menu is not configured to be displayed always, perhaps what we need to do is instead output two separate nav menus, one which doesn't have the `dialog` wrapper which is only displayed on non-mobile, and another which _is_ wrapped in `dialog` which can only be triggered to be displayed via the nav menu button which is only displayed on mobile. Only one would be displayed at a time, of course. Then only for the nav menu copy which is inside the `dialog` element would we then add the `autofocus` attribute to the first nav menu item.

stokesman commented:
> Would this be confusing from an accessibility perspective to have an element with a `dialog` role which is traversable/shown by default?

I think this is a key question to answer and I had meant to raise it too. It seems that it might be okay given that a dialog can be set to show statically by having the `open` attribute set.

> output two separate nav menus, one which doesn't have the `dialog` wrapper which is only displayed on non-mobile, and another which _is_ wrapped in `dialog` which can only be triggered to be displayed via the nav menu button which is only displayed on mobile

Alternatively we could leave focus to be handled by the UA unconditionally. Unless it’s determined that a `dialog` wrapper should be avoided for contexts where the navigation is not behind the toggle in which case I think there’s no other option besides outputting two separate menu contents.

stokesman committed:
Add open/close handlers only if Invoker commands aren’t supported

Also refactor to:
- minimize Interactivity state
- separate code paths used by submenus and overlay
- rename context’s `modal` to `menu` for clarity

stokesman committed:
Fix focus return for Escape key presses

Makes returning focus a responsibility of `closeMenu` so it applies
for both `handleMenuFocusout` and `handleMenuKey`.

stokesman committed:
Merge branch 'trunk' into try/navigation-block-dialog

stokesman committed:
Fixup of merge commit

westonruter reviewed (changes requested):
Just marking this as changes-requested since we still need to figure out the autofocus behavior when the nav menu is opened with an invoker command: https://github.com/WordPress/gutenberg/pull/69720#issuecomment-2773267709

If we change the JS to focus on the close button instead, then we don't need to add an `autofocus` attribute on the first focusable nav menu item since the browser will handle this by default. However, if for a11y reasons it is preferable to focus on the first nav menu item, then it seems we'll need to clone the nav menu specifically for the `DIALOG` for use on mobile, with the non-mobile nav menu not needing to be wrapped in a `DIALOG`. And if we do that, then we'll need to finish the HTML Tag Processor implementation to find additional focusable elements: https://github.com/WordPress/gutenberg/pull/69720/files#r2023568298

adamsilverstein commented on test/e2e/specs/editor/blocks/navigation-colors.spec.js:
```diff
@@ -271,6 +271,63 @@ test.describe( 'Navigation colors', () => {
 		await colorControl.testFrontendColors( expectedNavigationColors );
 	} );
 
+	test( 'The submenu/overlay colors do not apply to all navigation links unless the overlay is open', async ( {
```
these changes seem unrelated? (thru line 330)

stokesman replied:
Yes these are not directly related. This adds coverage for something [I broke while trying to clean up some styles](https://github.com/WordPress/gutenberg/pull/69720/commits/f550145e201402ccc95704372c6947ac60e29346#r2021845945). It’s possible that one of the existing tests could be modified so it would fail instead of adding this new one but I figured this is more explicit.

adamsilverstein commented on test/e2e/specs/editor/blocks/navigation-frontend-interactivity.spec.js:
```diff
@@ -65,67 +65,22 @@ test.describe( 'Navigation block - Frontend interactivity', () => {
-		 * remove this test and only rely on the Overlay Interactions test above and add a (@firefox, @webkit)
-		 * directive to the describe() statement. https://github.com/WordPress/gutenberg/pull/55198
-		 */
-		test( 'Overlay menu interactions in Safari (@webkit)', async ( {
```
Does this test no longer work or apply?

stokesman replied:
This test duplicates the test above it. As the comment preceding it indicates, the intention was to avoid flakiness of the test in Safari purportedly caused by Tab key presses. I removed those parts of the test so I followed up with the rest of what the comment recommends.

The reason I removed the assertions for Tab key and focus is because the “focus trapping“ implementation is removed and testing for it is not needed. The “focus trapping” can be removed because the `dialog` element, as a modal, makes all content outside it inert. Edit: also [commented here](https://github.com/WordPress/gutenberg/pull/69720#discussion_r2020407558)

Aside: this test was a complete duplicate of the one above it and included Tab key presses so I don’t see how it was actually making any difference for Safari in the first place.

westonruter commented:
Something that just came to mind: Chrome and Edge both support invoker commands, and they are  also implemented behind flags in Safari and Firefox ([caniuse](https://caniuse.com/mdn-html_elements_button_commandfor)). Once invoker commands ship in stable versions of Safari and Firefox, should we even keep the Interactivity API implementation here? We could instead ship the [invoker commands polyfill](https://github.com/keithamus/invokers-polyfill), if the current [a11y limitation](https://github.com/keithamus/invokers-polyfill?tab=readme-ov-file#limitations) is resolved. There is precedence for this in core as a polyfill for import maps was added in Core-60348. It was removed in Core-60970 after it was determined that all browsers support import maps. We could similarly have a `wp-load-polyfill-invokercommands` which the navigation block enqueues by default (or which any other block could enqueue if they use invoker commands), or even core could enqueue this automatically if the tag processor detects the `commandfor` attribute in any tag. 

Removing the Interactivity API implementation would reduce maintenance and it would reduce script from being served to the browser, thus improving performance. 

It's still too early here to do this, as there are other aspects of the Interactivity API which aren't yet able to be replaced with invokers, namely the dropdown menus. These rely on browsers implementing [interest invokers](https://open-ui.org/components/interest-invokers.explainer/). These are just now implemented in an [origin trial in Chrome 135](https://developer.chrome.com/release-notes/135#interest_invokers), and the PR to add this to the HTML spec is still just a draft. 

In any case, I just wanted to look ahead for what this implementation could eventually look like.

cc @swissspidy
